src/coreclr/System.Private.CoreLib/src/Internal/Runtime/InteropServices/ComActivationContextInternal.cs

```diff
@@ -16,7 +16,7 @@ public unsafe struct ComActivationContextInternal
         public char* AssemblyPathBuffer;
         public char* AssemblyNameBuffer;
         public char* TypeNameBuffer;
-        public IntPtr ClassFactoryDest;
+        public nint ClassFactoryDest;
```

Comment:
This and many other places are pointers. `nint` makes them look odd. We should be changing these to `void*` instead.

Reply:
I'm happy to do that, but just noting its going to make this a much more in depth change.

Also noting that this isn't critical for .NET 7, its mostly "stylistic" cleanup outside the places where `IntPtr.Zero` and similar were swapped to something more efficient.

Reply:
> I'm happy to do that, but just noting its going to make this a much more in depth change.

That's sounds fine to me. It can be done in chunks to make it reviewable.

I would switch to `void*` or other appropriate pointer type where possible, and keep using `IntPtr` for pointers where it is not possible to switch. Switch to nint/nuint only in places where the value is not pointer.

Reply:
> and keep using IntPtr for pointers where it is not possible to switch

This is going to conflict, longer term, with editorconfig options and what IntelliSense/docs display. As of C# 11 (when targeting .NET 7 or later), `IntPtr/nint` and `UIntPtr/nuint` are true aliases of eachother, in the same way as `Int32/int` and `UInt32/uint`. They are identical in IL (not even a `NativeInteger` attribute any longer) and all the same style rules and support will exist/apply.

Reply:
In my opinion it is about intent.  Yes, in IL they are identical aliases for each other. But the important thing here is for the code reader, not the compiler or tools.

What is the intent of this variable? Is it "just a number" like a size or a counter? Or is it a pointer?

I find it super confusing to use `nint` for "pointer". As a reader, you need to inspect further to figure out what the intention of the variable is than just looking at its type.

Reply:
There are two main things here:

1. The IDE will have no way to disambiguate and so it doesn't matter what we put in source. `nint` will be displayed by the tooling. These types in ECMA-335 (runtime spec) were never designed to be different. C# decided to disambiguate them while other languages, such as F# did not. C# trying to disambiguate them turned this into a massive leaky abstraction where its trying to express a difference that doesn't actually exist, which definitely doesn't exist in several other languages, and which fundamentally breaks down the more `System.IntPtr` gets used and gets support for integer like operations.

2. People having a pre-conceived notion that `IntPtr` means "pointer like" is wrong and needs to be changed. This is and always has been an incorrect association and is one that extends beyond the boundaries of just .NET. `IntPtr` comes from C/C++ where the corresponding name `intptr_t` exists. It is in the same format and of the same meaning as `Int32`. That is `Int` meaning `integer` and `32` meaning `32-bits`. Therefore `IntPtr` is `Int` meaning `integer` and `ptr` meaning `ptr-bits` (same width as a pointer).

The official guidance and what the `System.IntPtr` docs are being updated to indicate is that while you can use `IntPtr` to expose/represent/pass pointers and handles, this is unsafe and error prone. Usage needs to be carefully considered and primarily done in cases where the corresponding language doesn't have pointer support. In languages that support them handles and pointers should be represented as pointers or a type-safe wrapper (such as by correctly deriving `SafeHandle`). Trying to represent a `ushort*` as an `IntPtr` is unsafe, as is trying to represent a `HWND` as an `IntPtr`. Sometimes this is appropriate for interop code, but even there it is "technically incorrect" and could run into ABI issues on future platforms.

We should definitely follow our own guidance where it isn't a breaking change. But otherwise, the few remaining cases where it is a pointer being exposed "safely", people will get adjusted to the difference with time and it will be ultimately no different than seeing `intptr_t`, `uintptr_t`, or one of the many other typedefs in C/C++ which have the same meaning.

Reply:
Docs will have to fill the gap here and if there is a remaining concern, we always have the option of an analyzer or soft obsoletion and alternative overload for the small set of legacy APIs where `void*` should be exposed, and preferred, instead.

Reply:
> The IDE will have no way to disambiguate and so it doesn't matter what we put in source.

It matters very much to the readability of the code.

Reply:
Its likewise worth noting that an expectation of "pointer represented as an integer" is already incorrect due to a plethora of APIs exposed before `nint` existed in C# where `IntPtr` was literally a size.

`Marshal.AllocHGlobal` is probably the simplest example where it returns an `IntPtr` (a "safe pointer") and takes an `IntPtr` (a count representing the number of bytes to allocate).

This is purely preconceptions about what the two mean which are incorrect no matter which point in .NET history you look at. One cannot simply see `IntPtr` or `nint` and assume that it means `pointer` or `number`. The type must be used in coordination with the documentation, API names, and some amount of intuition.

Reply:
As you have said, this is only "stylistic" cleanup. C# does not have one universally accepted style. We are intentionally ignoring many style rules in this repo already because we believe that they would not improve the code readability. Nothing says that we have to enable style rule that says use nint/uint everywhere.

> Marshal.AllocHGlobal is probably the simplest example where it returns an IntPtr 

Yes, we cannot really fix these APIs. However, we can limit the readability damage. If there is a code that looks like this `IntPtr p = Marshal.AllocHGlobal(...);`, we should clean it up to be `void* p = (void*)Marshal.AllocHGlobal(...);`

Reply:
> C# does not have one universally accepted style.

Yes. But IntelliSense does and IntelliSense is going to display `nint` everywhere, just as it displays `int` everywhere.

IntelliSense "could" respect https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/style-rules/ide0049, but that would be `IntPtr` and `Int32` everywhere with `nint` and `int` nowhere.

> Nothing says that we have to enable style rule that says use nint/uint everywhere.

https://github.com/dotnet/runtime/blob/main/docs/coding-guidelines/coding-style.md. 

11. We use language keywords instead of BCL types (e.g. int, string, float instead of Int32, String, Single, etc) for both type references as well as method calls (e.g. int.Parse instead of Int32.Parse). See issue https://github.com/dotnet/runtime/issues/13976 for examples.

Now that `IntPtr` and `nint` are true aliases of eachother, `nint` is a keyword and is per our rules and the general styling/editorconfig rules, preferred.

-------------------------------------

Users are going to have to update their incorrect expectations here and change. This is and has always been the case for many languages outside of C#.

We can improve some things on our own end by utilizing pointers instead, where feasible. But trying to continue perpetuating the leaky abstraction that `IntPtr` is somehow different from `nint` is flawed and not going to end up in a positive area.

This was already discussed in depth by API review and C# LDM with consideration for how other languages do things, how APIs are exposed, and what our official guidance for exposing such APIs is.

Reply:
Whatever precedence we set in our own code is one that, for better or worse, is going to be used as a representative example for other codebases.

The handful of APIs that fall into the weird category of "this should have been `void*` but the API review committee at the time had an aversion to unsafe code" will be exposed to the end user as `nint` and that is how everyone is going to see that API in VS.

Reply:
> Whatever precedence we set in our own code is one that, for better or worse, is going to be used as a representative example for other codebases.

Agree. I believe that blank find&replace of IntPtr with nint is a bad precedence to set. We should set the precedence by converting IntPtr pointer values to use proper unmanaged pointer values where possible.

As the language evolves, the coding style has to evolve as well. I am proposing to add a new rule to the repo style rule: 
> We use unmanaged pointers (e.g. `void*`, `byte*`) for pointer values where possible. `IntPtr`/`nint` pointer values used by legacy APIs should be converted to and from unmanaged pointers at once (e.g. `void* p = (void*)Marshal.AllocHGlobal(...); Marshal.FreeGlobal((IntPtr)p);`).

Reply:
Right, I agree with that. The remaining bit is whether `AllocHGlobal` is declared in source as `IntPtr AllocHGlobal(nint cb)` or as `nint AllocHGlobal(nint cb)`.

IntelliSense, tooling, and other bits are all going to show and default to the latter for .NET 7+ and so I think we should be consistent there. The proposed rule you gave should be that we convert that to `void*`.

It might be beneficial to have an analyzer to help flag these edge cases, but that's not a strict need.

Reply:
> The remaining bit is whether AllocHGlobal is declared in source as IntPtr AllocHGlobal(nint cb) or as nint AllocHGlobal(nint cb).

I do not have a strong opinion on that. It is going to need a comment that explains that it is actually a pointer. It can be also declared as `nint /* IntPtr */ AllocHGlobal(nint cb)` or ``nint /* void* */ AllocHGlobal(nint cb)` to get the point across.